docs/posts/ibisml/index.qmd

```diff
@@ -0,0 +1,897 @@
+:::
+
+#### Aggregate features
+
+For tables with a depth greater than 0 that can't be directly joined with the base table,
+we need to aggregate the features by the `case_id`. You could compute the minimum,
+maximum, mean, median, and standard deviation for numeric columns, and the maximum and
+minimum for non-numeric columns.
+
+Here, we use the maximum as an example.
+
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to aggregate features by case_id using max"
+def agg_by_id(table):
+    return table.group_by("case_id").agg(
+        [
+            table[col_name].max().name(f"max_{col_name}")
+            for col_name in table.columns
+            if col_name[-1] in ("T", "L", "P", "A", "D", "M")
+        ]
+    )
+```
+::: {.callout-tip}
+For better predicting power, you need to collect different statistics based on the meaning of features. For simplicity, we'll only collect the maximum value of the features here.
+:::
+
+#### Read and process the data files
+
+We will define a function reads parquet files, optionally handles regex patterns for
+multiple sub-files, applies data type transformations using `data_type_recipes`, and
+performs aggregation based on `case_id` if specified by the depth parameter.
+
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to read and process data files"
+def read_and_process_files(file_path, depth=None, is_regex=False):
+    """
+    Read and process Parquet files.
+
+    Args:
+        file_path (str): Path to the file or regex pattern to match files.
+        depth (int, optional): Depth of processing. If 1 or 2, additional aggregation is performed.
+        is_regex (bool, optional): Whether the file_path is a regex pattern.
+
+    Returns:
+        ibis.Table: The processed Ibis table.
+    """
+    if is_regex:
+        # Read and union multiple files
+        chunks = []
+        for path in glob(str(file_path)):
+            chunk = ibis.read_parquet(path)
+            # Transform table using IbisML Recipe
+            chunk = data_type_recipes.fit(chunk).to_ibis(chunk)
+            chunks.append(chunk)
+        table = ibis.union(*chunks)
+    else:
+        # Read a single file
+        table = ibis.read_parquet(file_path)
+        # Transform table using IbisML
+        table = data_type_recipes.fit(table).to_ibis(table)
+
+    # Perform aggregation if depth is 1 or 2
+    if depth in [1, 2]:
+        table = agg_by_id(table)
+
+    return table
+```
+
+Let's define two dictionaries, `train_data_store` and `test_data_store`, that organize and
+store processed datasets for training and testing.
+
+```{python}
+#| code-fold: true
+#| code-summary: "Show code to load all data into a dict"
+train_data_store = {
+    "df_base": read_and_process_files(TRAIN_DIR / "train_base.parquet"),
+    "depth_0": [
+        read_and_process_files(TRAIN_DIR / "train_static_cb_0.parquet"),
+        read_and_process_files(TRAIN_DIR / "train_static_0_*.parquet", is_regex=True),
```

Comment:
Nit: Might consider some list comprehensions in this section. :D A lot of repetition.